ucp/endpoint_reuse.py

```diff
@@ -0,0 +1,154 @@
+import asyncio
+import uuid
+
+import numpy as np
+
+from . import core
+
+
+class EPHandle:
+    def __init__(self, ep):
+        self.ep = ep
+        self.refcount = 1
+
+
+class EndpointReuse:
+    """
+    Class to seamlessly reuse endpoints. It uses the tag feature of send/recv
+    to separate "virtual" endpoint pairs from each other.
+
+    Connection Protocol
+    -------------------
+    1) Client connect to server using a new endpoint.
+    2) Client send the IDs of all its existing endpoints.
+    3) Server receives the IDs from the client and checks if it has a
+       matching existing endpoint. It then sends the matching ID to the
+       client or zero if no match.
+    4) The client and server now continue with either the existing matching
+       endpoints or the new endpoints (which are registered for later reuse).
+    """
+
+    existing_endpoints = {}
+
+    def __init__(self, handle, tag, close_tag):
+        self.handle = handle
+        self.tag = tag
+        self.close_tag = close_tag
+        self.close_recv = asyncio.create_task(
+            handle.ep.recv(bytearray(1), tag=close_tag)
+        )
+
+    @classmethod
+    async def create_endpoint(cls, ip, port):
+        tag = np.uint32(uuid.uuid4().int % np.iinfo(np.uint32).max)
+        close_tag = np.uint32(uuid.uuid4().int % np.iinfo(np.uint32).max)
```

Comment:
As to the question raised earlier about `iinfo`, I think we can just cast these and it will only keep the relevant "lower" bits. IOW `iinfo` shouldn't be needed.

```python
In [1]: import ctypes, uuid                                                     

In [2]: i = uuid.uuid4().int; i                                                 
Out[2]: 287371484024313512047830668633417152082

In [3]: i % 2**32                                                               
Out[3]: 692024914

In [4]: ctypes.c_uint32(i)                                                      
Out[4]: c_uint(692024914)
```

```suggestion
        tag = ctypes.c_uint32(uuid.uuid4().int)
        close_tag = ctypes.c_uint32(uuid.uuid4().int)
```

Reply:
Thanks @jakirkham, added in https://github.com/rapidsai/ucx-py/pull/556/commits/b3b3dcd16049ec6c22dda6d9b98706aa418779dc